client/src/main/scala/org/http4s/client/middleware/GZip.scala

```diff
@@ -37,25 +40,37 @@ object GZip {
           req.headers ++ Headers.of(Header(`Accept-Encoding`.name.value, supportedCompressions)))
     }
 
-  private def decompress[F[_]](bufferSize: Int, response: Response[F])(implicit
-      F: Sync[F]): Response[F] =
+  private def canEntityBodyBeEmpty(requestMethod: Method): Boolean =
```

Comment:
Yeah, I totally agree. Actually, I started fixing the issue by trying to consume the first chunk from the stream and see if it's empty, but I run into another issue (sometimes the `java.util.zip.Inflater` used in the `fs2.compression.gunzip` method was being closed before the unzip process finished and it crashed; probably I did something wrong, I have to dive deeper to understand why that happened, I'm not very familliar with the `fs2` codebase).
Maybe we could just remove the verification that the body can be empty (if expression on line 73, Gzip.scala) and always handle `EOFException` by returning an empty stream, as that would cover all cases?

Reply:
What happens if the gzipped stream is truncated?  Would that raise an EOFException?  If so, I think that's one that we'd want to leave raised.

Reply:
I've taken a closer look at `fs2.compression.gunzip` method and the `EOFException` is used three times: initially checking if the mandatory header is present, then when trying to read the optional external field which is indicated to exist by the byte of flags, or when trying to go farther reading the filename, comment or content. And these cases, indeed, represent a truncated stream.

I would be glad to work on this farther, either on this PR or in another one, by initially inspecting whether the stream is empty before proceeding to decompression, as you suggested. I would need, probably, a bit more time for it, because I have time to work on this only on weekends; if that is okay.

Reply:
I think the current form fixes the HEAD issue without making anything wors.  Let's go ahead and merge this, and we can handle these other cases in a subsequent PR.

Reply:
Ok, thanks. Then, I hope to open soon a subsequent PR to cover the other cases.